tc/core/tc_executor_api_rfc.h

```diff
@@ -0,0 +1,138 @@
+ * run on a specified backend. The correspondence is 1 TcExecutor for 1
+ *   tuple<TC function, input shapes, MappingOptions>.
+ *
+ * TcExecutor is an abstract base class which provides capabilities to:
+ *   1. compile a TC function into a CompilationResult and an executable
+ *      RTCFunction;
+ *   2. store an abstraction of ExecutionContext and RTCFunction to check and
+ *      run the result of step 1. multiple times on different tensors of
+ *      compliant sizes;
+ *   4. interact with the compilation CacheType to allow saving and loading
+ *      the results of step 1. across executions. The TC function
+ *      representation in the cache is a lang::CanonicalTcString.
+ *      The CompilationResult is exactly a (protobuf-backed) cache entry.
+ *
+ * TcExecutor is templated by the Backend type.
+ * For each backend, the specific Backend type lives in
+ *   core/backend/backend_tc_executor.h and declares all the required dependent
+ *   **concrete** types.
+ * For example:
+ *   CudaBackend is declared in core/cuda/cuda_tc_executor.h
+ *
+ * struct CudaBackend {
+ *   using ExecutorType = CudaTcExecutor;
+ *   using MappingOptionsType = CudaMappingOptions;
+ *   using ProtobufType = CudaCacheProto;
+ *   using CacheType = CudaCache;
+ *   using CacheEntryType = CudaCacheEntry;
+ *   using CacheKeyType = typename CacheEntryType::Key;
+ *   using CompilationResultType = typename CacheEntryType::Values;
+ *   using RTCFunctionType = CudaRTCFunction;
+ * };
+ *
+ * Backend-specific Executors should specialize TcExecutor<Backend> and
+ * implement the uncheckedRun and setRuntimeCompiledFunction.
+ */
+template <typename Backend>
+class TcExecutor {
+ protected:
+  TcExecutor(
+      lang::TreeRef tcDefinition,
+      const std::vector<const DLConstTensor*>& inputs);
+
+ public:
+  virtual ~TcExecutor();
+
+  TcExecutor(TcExecutor&&) = delete;
+  TcExecutor& operator=(TcExecutor&&) = delete;
+  TcExecutor(const TcExecutor&) = delete;
+  TcExecutor& operator=(const TcExecutor&) = delete;
+
+  /// Run can be called multiple times given a compilation, inputs are allowed
+  /// to change in that their data pointer is allowed to change.
+  /// Sizes and strides must remain constant otherwise this is an error.
+  /// The only thing that is allowed to change across runs is the input
+  /// and output pointers base address.
+  /// It is the caller's responsibility to ensure proper non-aliasing (or
+  /// advanced aliasing) properties of the input and output tensors.
+  /// if profile is set the kernel runtime (microseconds) is returned.
+  /// \returns profiling information (see: tc/core/utils/time.h)
+  ProfilingInfo run(
+      const std::vector<const DLConstTensor*>& inputs,
+      const std::vector<const DLTensor*>& outputs,
+      bool profileKernel = false) const;
+
+  /// It is necessary to clear the RTC manually because it can throw and we
+  /// can't have that in the destructor.
+  inline void clearRuntimeCompiledFunction();
```

Comment:
If `setupRuntimeCompiledFunction` is virtual, so should be `clearRuntimeCompiledFunction`.  It may be concrete and empty though.